### PR TITLE
[EPLB]: Optimize `export_load_view` update

### PR DESCRIPTION
**Pull request by dragondream-chen**
## Purpose
In the feature of EPLB(Experts Load Balance), the PR optimizes the update method for expert load during each forward. The current approach is using the `scatter_add_` method based on `topk_ids` results. When using DeepEP Low-Latency or PPLX on the CUDA platform, expert loads can be obtained directly from `expert_tokens_meta.expert_num_tokens`, which reduces redundant calculations on the expert load. 

## Test Plan
1. Test expert load update 
Since the use of the kernel, such as DeepEP Low-Latency or PPLX, leads to some changes in the inference process, the precision of intermediate values cannot be fully aligned. We directly illustrate whether the expert load update function is functioning properly by comparing the load imbalance of one layer model.
We add the following code in `vllm/distributed/eplb/eplb_state.py` for data collection. 
```python
if global_expert_load is None:
     physical_expert_load_window = self.expert_load_window.clone()
     global_physical_load_window = physical_expert_load_window.sum(dim=0)
     all_reduce(global_physical_load_window, group=ep_group)
if is_main_rank:
    global_num_experts = 96
    ep_size = ep_group.size()
    all_rank_node = []
    for ep_r in range(ep_size):
        base_experts = global_num_experts // ep_size
        remainder = global_num_experts % ep_size
        if er < remainder:
            local_num_experts = base_experts + 1
        else:
            local_num_experts = base_experts
        # Create a tensor of size num_experts filled with -1
        expert_map = torch.zeros(global_num_experts, device=global_physical_load_window.device, dtype=global_physical_load_window.dtype)
        # Create a expert map for the local experts
        start_idx = ep_r * base_experts + min(ep_r, remainder)
        expert_map[start_idx:start_idx + local_num_experts] = 1
        
        # [layers, phy_num] * [phy_num,]
        local_load = (global_physical_load_window * expert_map).sum(1).unsqueeze(1)
        all_rank_node.append(local_load)
    all_ranks = torch.cat(all_rank_node, dim=1).float() # [layers, ep_size]: [26, 8]
    max_ranks, _ = torch.max(all_ranks, dim=1) # [layers]
    mean_ranks = torch.mean(all_ranks, dim=1) # [layers]
    unbanlance = torch.div(max_ranks, mean_ranks) # [layers]
    logger.debug(f" | unbanlance : {unbanlance.cpu().tolist()}")
    for i in range(len(unbanlance)):
        logger.debug(f" | unbanlance layer {i} : {unbanlance[i]} ")
```

2. Test  performance
Test the average time for a single update of the expert. 

## Test Result
image
The blue curve represents the state before modification, while the red curve represents the state after modification. 
From the graph, it can be observed that the degree of imbalance is essentially similar. 

The average time of update `expert_load_view` for one layer in every batch:
Before modification: **0.058ms**
After modification: **0.015ms**

# Co-author
Co-authored-by: Skywalker-EP 173723846@qq.com

### CODE REVIEW

**Comment by dragondream-chen** [comment]
Hi @simon-mo and @khluu,

I just submitted my first PR to VLLM. I don’t have permission to unblock additional CI tests on Buildkite (only fastcheck runs by default). Could you help add me to the vLLM Buildkite org so I can trigger full CI?

Thanks for your help!

**Comment by abmfy** [comment]
Since #22167, we have been collecting the global physical expert loads, so the overhead of masking out non-local experts has already been removed.

In addition, please see #24573, where the unnecessary local mask was removed. After that change, the only remaining operation for expert load collection is a `scatter_add_`, which is essentially the same as what this PR proposes.

**Comment by dragondream-chen** [comment]
> Since #22167, we have been collecting the global physical expert loads, so the overhead of masking out non-local experts has already been removed.
> 
> In addition, please see #24573, where the unnecessary local mask was removed. After that change, the only remaining operation for expert load collection is a `scatter_add_`, which is essentially the same as what this PR proposes.

PR 24573 does not conflict with our optimization for the `expert_load_view` update. This PR aligns with SGLang's approach and addresses the TODO of `expert_load_view`. 
We directly updates the `expert_load_view` using `expert_num_tokens` obtained from `FusedMoEPrepareAndFinalize`. This method can reduce some additional processing on `topk_ids`, and the dimension handled by `scatter_add_` is reduced from `batch_size x expert_num` to `physical_expert_num`. 
Further discussion on this would be greatly appreciated!

**Comment by abmfy** [comment]
Sorry for the delay.  

I see your point. You're right that previously we needed to perform `scatter_add_` for `batch_size * topk` times, whereas now it's just a simple tensor addition over `num_local_experts`, which should indeed be faster.  

To proceed, I think we should do the following:  

1. **Make it more general.** This PR currently hardcodes `DeepEPHTPrepareAndFinalize` to use this simplified path, but in fact `DeepLLPrepareAndFinalize`, `BatchedPrepareAndFinalize`, and `PplxPrepareAndFinalize` all provide expert token count metadata. We can add a method in the ABC `mk.FusedMoEPrepareAndFinalize` that indicates whether a prepare-and-finalize kernel returns expert token count metadata. By default, it can return `False`, and we can override it to return `True` in the aforementioned classes.  

2. **Simplify token count addition.** The token count in `expert_tokens_metadata` maps to local experts, so we can perform a direct tensor addition instead of `scatter_add_`. Specifically, use integer indexing like:  
   ```python
   expert_load_view[physical_expert_start:physical_expert_end] += expert_tokens_metadata.expert_num_tokens
   ```  
   - The multiple abstraction layers (`FusedMoE` → `QuantMethod` → `FusedMoEModularKernel`) make it tricky to pass information directly. If you can find a better way to pass start/end indexes directly, that would be preferable to inferring them from `expert_map` every time.  
   - Avoid using `torch.arange` for indexing here, as that effectively becomes a `scatter_add_` again and would be slower than simple indexing.  

3. **Benchmarking.** Provide benchmark results. Make sure to lower `eplb_config.step_interval` and test on large datasets like ShareGPT (not just a few prompts) to properly trigger EPLB, e.g., run for 100 steps.  

4. **Accuracy test.** Test GSM8k before and after this PR to verify correctness.  

Thanks again for your contribution!

**Comment by hmellor** [comment]
In future please take more care with git and force push. Almost all maintainers are now subcribed to this PR because of it.

Also, please:

- Fix pre-commit
  ```bash
  uv pip install pre-commit
  pre-commit install
  pre-commit run -a
  ```
- Merge from main to fix the doc issue

**Comment by Mercykid-bash** [comment]
For the token count addition simplification, we've replaced `scatter_add_` with direct tensor addition. Here, `physical_expert_start` and `physical_expert_end` are now derived using the rank ID and local expert count, avoiding inference from `expert_map` or `torch.arange` usage.

Initial tests have been completed to verify correctness. We'll follow up with the Benchmarking and Accuracy test experiments shortly.

**Comment by dsxsteven** [comment]
> Sorry for the delay.
> 
> I see your point. You're right that previously we needed to perform `scatter_add_` for `batch_size * topk` times, whereas now it's just a simple tensor addition over `num_local_experts`, which should indeed be faster.
> 
> To proceed, I think we should do the following:
> 
> 1. **Make it more general.** This PR currently hardcodes `DeepEPHTPrepareAndFinalize` to use this simplified path, but in fact `DeepLLPrepareAndFinalize`, `BatchedPrepareAndFinalize`, and `PplxPrepareAndFinalize` all provide expert token count metadata. We can add a method in the ABC `mk.FusedMoEPrepareAndFinalize` that indicates whether a prepare-and-finalize kernel returns expert token count metadata. By default, it can return `False`, and we can override it to return `True` in the aforementioned classes.
> 2. **Simplify token count addition.** The token count in `expert_tokens_metadata` maps to local experts, so we can perform a direct tensor addition instead of `scatter_add_`. Specifically, use integer indexing like:
>    ```python
>    expert_load_view[physical_expert_start:physical_expert_end] += expert_tokens_metadata.expert_num_tokens
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    
>    * The multiple abstraction layers (`FusedMoE` → `QuantMethod` → `FusedMoEModularKernel`) make it tricky to pass information directly. If you can find a better way to pass start/end indexes directly, that would be preferable to inferring them from `expert_map` every time.
>    * Avoid using `torch.arange` for indexing here, as that effectively becomes a `scatter_add_` again and would be slower than simple indexing.
> 3. **Benchmarking.** Provide benchmark results. Make sure to lower `eplb_config.step_interval` and test on large datasets like ShareGPT (not just a few prompts) to properly trigger EPLB, e.g., run for 100 steps.
> 4. **Accuracy test.** Test GSM8k before and after this PR to verify correctness.
> 
> Thanks again for your contribution!

Hi, I used the Deepseek-V2-Lite model and gsm8k dataset for performance testing. However, after enabling PPLX, there were accuracy issues. Before and after enabling PPLX, the accuracy was only around 2%. Below are the results from my AISbench test.

results before PR
image

results after PR
image

**Comment by dsxsteven** [comment]
I used the GSM8k dataset for performance testing and lowered "window_size":1000,"step_interval":300 to trigger EPLB.
As you can see, after enabling our pull request, **both TTFT and TPOT decreased**, resulting in a performance gain.

performance before PR
image

performance after PR
image
